src/twisted/internet/tcp.py

```diff
@@ -853,6 +858,246 @@ def getPeer(self):
 
 
 
+def _silenceOutputs(outputs):
+    """
+    An L{automat} output collector that returns L{None}.
+
+    @param outputs: An outputs iterator
+
+    @return: L{None}
+    """
+    list(outputs)
+    return None
+
+
+
+def _nthOutput(n):
+    """
+    An L{automat} output collector factory that returns the value of
+    the C{nth} output.
+
+    @param n: An integer representing C{nth} output to return.
+        Counting starts at zero.
+
+    @return: The value of the first output.
+    """
+    def _collectNthOutput(outputs):
+        return list(outputs)[n]
+    return _collectNthOutput
+
+
+
+@attr.s
+class _FileDescriptorReservation(object):
+    """
+    An open file that represents an emergency reservation in the
+    process' file descriptor table.  If L{Port} encounters C{EMFILE}
+    on C{accept(2)}, it can close this file descriptor, retry the
+    C{accept} so that the incoming connection occupies this file
+    descriptor's space, and then close that connection and reopen this
+    one.
+
+    Calling L{_FileDescriptorReservation.acquire} opens the reserve
+    file descriptor if it is not already open.
+    L{_FileDescriptorReservation.release} close this descriptor; a
+    subsequent call to L{_FileDescriptorReservation.replace} will
+    close the provided file and re-open the reservation file.  A
+    claimed reservation will not replace the provided file with
+    itself.  L{_FileDescriptorReservation.release} is idempotent.
+
+    @param fileFactory: A factory that will be called to reserve a
+        file descriptor.
+    @type fileFactory: A L{callable} that accepts no arguments and
+        returns an object with a C{close} method.
+    """
+    _log = Logger()
+    _machine = automat.MethodicalMachine()
+
+    _fileFactory = attr.ib()
+    _fileDescriptor = attr.ib(init=False, default=None)
+
+    @_machine.state(initial=True)
+    def _unreserved(self):
+        """
+        The reservation file is unreserved.
+        """
+
+
+    @_machine.state()
+    def _reserved(self):
+        """
+        The reservation file is reserved.
+        """
+
+
+    @_machine.input()
+    def available(self):
+        """
+        Is the reservation available?
+        """
+
+
+    @_machine.input()
+    def release(self):
+        """
+        Release the reservation descriptor so another file can take
+        its place.
+        """
+
+
+    @_machine.input()
+    def acquire(self):
+        """
+        Ensure that the reservation file is held.
+        """
+
+
+    @_machine.input()
+    def replace(self, fileObject):
+        """
+        Replace a file with the reservation file only if our reserve
+        file is unclaimed.
+
+        @param fileObject: The file object to close.
+        """
```

Comment:
It really means: tell the caller if the provided file was closed because the reservation was made available by a previous call to `release`.  If it wasn't, then the reservation is still held because `EMFILE` has not occurred.

Naming is hard and I'm bad at it.  Maybe `maybeClose` is a better name?

Reply:
Addressed in: https://github.com/twisted/twisted/pull/996/commits/124437a0814e35c6dafddfa21ec2f337960af5e1